engine/rpcserver.go

```diff
@@ -1832,39 +1889,290 @@ func (s *RPCServer) GCTScriptListAll(context.Context, *gctrpc.GCTScriptListAllRe
+		RunningUrl:    w.GetWebsocketURL(),
+		ProxyAddress:  w.GetProxyAddress(),
+	}, nil
+}
+
+// WebsocketSetEnabled enables or disables the websocket client
+func (s *RPCServer) WebsocketSetEnabled(_ context.Context, r *gctrpc.WebsocketSetEnabledRequest) (*gctrpc.GenericResponse, error) {
+	exch := GetExchangeByName(r.Exchange)
+	if exch == nil {
+		return nil, errExchangeNotLoaded
+	}
+
+	w, err := exch.GetWebsocket()
+	if err != nil {
+		return nil, fmt.Errorf("websocket not supported for exchange %s", r.Exchange)
+	}
+
+	exchCfg, err := Bot.Config.GetExchangeConfig(r.Exchange)
+	if err != nil {
+		return nil, err
+	}
+
+	if r.Enable {
+		err = w.Enable()
+		if err != nil {
+			return nil, err
+		}
+
+		exchCfg.Features.Enabled.Websocket = true
+		return new(gctrpc.GenericResponse), nil
+	}
+
+	err = w.Disable()
+	if err != nil {
+		return nil, err
+	}
+	exchCfg.Features.Enabled.Websocket = false
+	return new(gctrpc.GenericResponse), nil
+}
+
+// WebsocketGetSubscriptions returns websocket subscription analysis
+func (s *RPCServer) WebsocketGetSubscriptions(_ context.Context, r *gctrpc.WebsocketGetSubscriptionsRequest) (*gctrpc.WebsocketGetSubscriptionsResponse, error) {
+	exch := GetExchangeByName(r.Exchange)
+	if exch == nil {
+		return nil, errExchangeNotLoaded
+	}
+
+	w, err := exch.GetWebsocket()
+	if err != nil {
+		return nil, fmt.Errorf("websocket not supported for exchange %s", r.Exchange)
+	}
+
+	payload := new(gctrpc.WebsocketGetSubscriptionsResponse)
+	payload.Exchange = exch.GetName()
+	subs := w.GetSubscriptions()
+	for i := range subs {
+		params, err := json.Marshal(subs[i].Params)
+		if err != nil {
+			return nil, err
+		}
+		payload.Subscriptions = append(payload.Subscriptions,
+			&gctrpc.WebsocketSubscription{
+				Channel:  subs[i].Channel,
+				Currency: subs[i].Currency.String(),
+				Asset:    subs[i].Asset.String(),
+				Params:   string(params),
+			})
+	}
+	return payload, nil
+}
+
+// WebsocketSetProxy sets client websocket connection proxy
+func (s *RPCServer) WebsocketSetProxy(_ context.Context, r *gctrpc.WebsocketSetProxyRequest) (*gctrpc.GenericResponse, error) {
+	exch := GetExchangeByName(r.Exchange)
+	if exch == nil {
+		return nil, errExchangeNotLoaded
+	}
+
+	w, err := exch.GetWebsocket()
+	if err != nil {
+		return nil, fmt.Errorf("websocket not supported for exchange %s", r.Exchange)
+	}
+
+	err = w.SetProxyAddress(r.Proxy)
+	if err != nil {
+		return nil, err
+	}
+	return new(gctrpc.GenericResponse), nil
+}
+
+// WebsocketSetURL sets exchange websocket client connection URL
+func (s *RPCServer) WebsocketSetURL(_ context.Context, r *gctrpc.WebsocketSetURLRequest) (*gctrpc.GenericResponse, error) {
+	exch := GetExchangeByName(r.Exchange)
+	if exch == nil {
+		return nil, errExchangeNotLoaded
+	}
+
+	w, err := exch.GetWebsocket()
+	if err != nil {
+		return nil, fmt.Errorf("websocket not supported for exchange %s", r.Exchange)
+	}
+
+	err = w.SetWebsocketURL(r.Url, false, true)
```

Comment:
I'm not getting a response/its hanging:
```
.\gctcli.exe websocket seturl --exchange=bitfinex --url=wss://echo.websocket.org
```

I am getting updates on GCT side though, so it looks like the set was successful
```
[DEBUG] | WEBSOCKET | 15/07/2020 14:11:15 | Bitfinex websocket: setting unauthenticated websocket URL: wss://echo.websocket.org
[DEBUG] | WEBSOCKET | 15/07/2020 14:11:15 | Bitfinex websocket: flushing websocket connection to wss://echo.websocket.org
[DEBUG] | WEBSOCKET | 15/07/2020 14:11:15 | Bitfinex websocket: shutting down websocket
[DEBUG] | WEBSOCKET | 15/07/2020 14:11:16 | Bitfinex websocket: running connection monitor cycle
```

Reply:
I had this too and discussed it with @shazbert but with using a dodgy wss url. Possibly may be due to the wss connection Dial'ing